Add `get_ref` methods for Queries

Adamkob12 opened the pull request:
# Objective

- fix #11517 by adding a `get_ref` method for `Query` and `QueryState`

## Solution

-  Add `get_ref` method for `QueryState`
-  Add `get_ref` method for `Query`

## Changelog

- The `QueryData` trait has another associated type - `ReadOnlySmartRef`. `ReadOnlySmartRef` is the type we get when we request read-only, change-detection-enabled, shared access to the data. For example:
access of `&T` is `Ref<T>`, the same is true for `&mut T`, because this access is explicitly read-only, and as such is analogous to first converting a `QueryData` to its read-only form (`QueryData::ReadOnly`) and then to its `QueryData::ReadOnlySmartRef` form. 
- The `QueryData` derive macro changed a bit in its implementation to reflect the above changes.
- The `QueryData` implementations of many types have changed to reflect the above changes.
- `get_ref` , `get_component_ref`, `iter_ref`, `as_readonly_smart_ref` methods were added to `QueryState`
- `get_ref`, `iter_ref`, `iter_combinations_ref`, `iter_many_ref`, `par_iter_ref`, `get_component_ref`, `component_ref`, `single_ref`, `get_single_ref`,  methods was added to `Query`
- `system_query_get_ref` test was added to make sure `Query<Ref<T>>` and `Query<&T>` + `get_ref` produce the same result. 

## Migration guide

- An associated type was added to the public `QueryData` trait: `ReadOnlySmartRef`. If the type you're implementing `QueryData` for can be accessed using some change-detection-enabled, read-only way, for example: the `Ref` smart pointer - then define `ReadOnlySmartRef` to be that type. If it can't be accessed that way, then keep `ReadOnlySmartRef` the same as `ReadOnly`.
 For example:
```rust
// Before:
impl<'a, C: Component> QueryData for &'a C {
    type ReadOnly = &'a C; // Regular read-only access
}

impl QueryData for Entity {
    type ReadOnly = Entity; // Standard read-only access of a Copy type
}

// After:
impl<'a, C: Component> QueryData for &'a C {
    type ReadOnly = &'a C; // Regular read-only access
    type ReadOnlySmartRef = Ref<'a, C>; // We can use `Ref` for change detection
}

impl QueryData for Entity {
    type ReadOnly = Entity; // Standard read-only access of a Copy type
    type ReadOnlySmartRef = Entity; // No option for change-detection, so keep it the same as `Self::ReadOnly`
}

```

alice-i-cecile commented on crates/bevy_ecs/src/query/fetch.rs:
```diff
@@ -276,6 +283,10 @@ pub unsafe trait ReadOnlyQueryData: QueryData<ReadOnly = Self> {}
 pub type QueryItem<'w, Q> = <Q as WorldQuery>::Item<'w>;
 /// The read-only variant of the item type returned when a [`QueryData`] is iterated over immutably
 pub type ROQueryItem<'w, D> = QueryItem<'w, <D as QueryData>::ReadOnly>;
+// TODO: Better docs, not sure about using the term "reffed".
```
I think we can probably use something like "smart-pointer" or "change-detection enabled read-only reference".

alice-i-cecile commented on crates/bevy_ecs/src/query/fetch.rs:
```diff
@@ -276,6 +283,10 @@ pub unsafe trait ReadOnlyQueryData: QueryData<ReadOnly = Self> {}
 pub type ROQueryItem<'w, D> = QueryItem<'w, <D as QueryData>::ReadOnly>;
+// TODO: Better docs, not sure about using the term "reffed".
+/// The reffed variant of the item type returned when a "reffed access" is requested. For example,
+/// Ref<T> would be "reffed access" of &T.
```
Remember to doc link!

alice-i-cecile commented on crates/bevy_ecs/src/query/state.rs:
```diff
@@ -67,6 +67,17 @@ impl<D: QueryData, F: QueryFilter> QueryState<D, F> {
+        // SAFETY: invariant on `WorldQuery` trait upholds that `D::Reffed` and `F::ReadOnly`
+        // have a subset of the access, and match the exact same archetypes/tables as `D`/`F` respectively.
+        //
+        // ------- WHICH I'M NOT SURE IS TRUE FOR &T and Ref<T> -------- (remove if PR is reviewed and accepted)
```
I believe this is true, yes.

alice-i-cecile commented on crates/bevy_ecs/src/query/state.rs:
```diff
@@ -398,6 +409,29 @@ impl<D: QueryData, F: QueryFilter> QueryState<D, F> {
     }
 
+    // TODO: Better explanation, I'm not a fan of the term "reffed".
+    /// Gets the query result in "reffed" form (whatever can go in the [`Ref`] smart pointer - goes into it,
```
Perhaps something like "Gets a read-only smart pointer with change-detection information to the query results for the given `World` and `Entity`."

alice-i-cecile commented on crates/bevy_ecs/src/query/state.rs:
```diff
@@ -398,6 +409,29 @@ impl<D: QueryData, F: QueryFilter> QueryState<D, F> {
+    /// Gets the query result in "reffed" form (whatever can go in the [`Ref`] smart pointer - goes into it,
+    /// whatver doesn't - turns into the read-only version) for the given [`World`] and [`Entity`].
+    ///
+    /// This can only be called for read-only queries, see [`Self::get_mut`] for write-queries.
```
I don't think this should be required: can we call `QueryState::get` on write-queries?

Adamkob12 replied:
I copied it from `QueryState::get`, I thought the same thing but wasn't sure.

Adamkob12 commented:
@alice-i-cecile What do I do about that one failing CI that expects a certain error message? (because the derive macro changed a bit and that messed with the error)

alice-i-cecile commented:
> @alice-i-cecile What do I do about that one failing CI that expects a certain error message? (because the derive macro changed a bit and that messed with the error)

The error message should be updated to match in this PR. IIRC there's a "bless" command to do it automatically if you prefer: check the docs of the crate we're using if you want to try it out :)

alice-i-cecile reviewed (changes requested):
Once we have a better name for `ReffedQueryItem` (and similar types/methods) I'm happy with this PR.

Something like "FatRef" or "ReadOnlyMetadata" or "SmartReadOnly" would work. I think `SmartReadOnly` is my preferred term, having chewed on the problem for a bit. It's fairly clear, and makes its relation to the other associated type apparent.

Remember to add a migration guide: we've added a new associated type to a public trait :)

viridia commented:
Naming suggestions:
* `QueryItemRef` (`Mut` is often used as a suffix, so why not `Ref`?)
* `ValueRef`
* `WithRef`
* `RefWrapper`
* `RefAccessor`
* `RefAccess`
* `Reference`

chrisjuchem reviewed (commented):
Is there a reason it's not possible for the "reffed" form of `&'a mut C` to be `Mut<'a, C>`? That would be my intuitive expectation, to just add change-detection without changing the rest of the semantics.

I additionally thing it would be nice if there was a way to get the reffed version of an entire query rather than just for a single entity, e.g. to iterate over all the reffed versions while still applying the same filters as some original, non-reffed query.

Also adding `WithChangeDetection`/`get_with_change_detection` as a name suggestion.

alice-i-cecile commented:
> Is there a reason it's not possible for the "reffed" form of &'a mut C to be Mut<'a, C>? That would be my intuitive expectation, to just add change-detection without changing the rest of the semantics.

The default form of `&'a mut C` already is `Mut<'a, C>` 🤔 I think that "the reffed form is always read-only" is more clear, and works better with the rest of the API.

Adamkob12 commented:
I settled on `ROQueryItemRef` and `ReadOnlySmartRef`, I think these are consistent and straightforward. I implemented many of your suggestions in the documentation, and added a migration guide.
Should I add the missing "get_ref API" such as `iter_ref` etc. here or in another PR? or not at all?

Adamkob12 commented on crates/bevy_ecs/src/query/fetch.rs:
```diff
@@ -263,6 +263,11 @@ use std::{cell::UnsafeCell, marker::PhantomData};
     type ReadOnly: ReadOnlyQueryData<State = <Self as WorldQuery>::State>;
+    /// The "reffed" variant of this [`QueryData`] (as in, the [`Ref`] smart pointer for change detection).
+    /// - If this [`QueryData`] doesn't have a "reffed" form (for example, [`Entity`]) - this type must be identical to [`Self::ReadOnly`].
+    /// - If this [`QueryData`] does have a "reffed" form (for example, &'w T -> Ref<'w, T>) then that would be this type.
```
I agree, the "missing API" (such as `iter_ref`, `get_single_ref`, etc.) should be completed, either here or in another PR.
I don't know if these should be added to the safety spec because the current one is specific enough.

alice-i-cecile reviewed (approved):
Great, I'm happy with this! I'd like to see `iter_ref` as well, but I have no strong feelings for this PR vs a follow-up one.

hymm commented:
I'm unsure of the utility of this PR. I feel like if you have a `Query<&T>` it's easy enough to change that to a `Query<Ref<T>>`. So the only time this comes in handy is when you have a `Query<&mut T>`. I feel like it would be easier to change the ReadOnly associated type to a `Ref<T>` for `impl WorldData for &mut T`. It would be a breaking change, but I think it makes sense for `Mut<T>` to go to a `Ref<T>` when it's read only instead of a `&T`. This would make `query.get` return a `Ref<T>` for `Query<&mut T>`.

james7132 commented:
I'm also going to voice my opposition to this, quite frankly, signifcant complexity of the implementation and how much utility it provides. Another option would be to allow `Mut<T>` to downgrade into a `Ref<T>` for a similar effect.

Adamkob12 commented:
I agree, I've felt this is too much for too little.
